src/bosh_azure_cpi/lib/cloud/azure/helpers.rb

```diff
@@ -544,232 +537,26 @@ def get_os_disk_size(root_disk_size, stemcell_info, use_root_disk)
-    #     # Please note, the mutex should NOT be unlocked, because other processes will end mutex.wait and continue to use the shared resource if unlocked.
-    #     # If your work is a long-running task,
-    #     #   you need to call mutex.update() in do_something() to update the lock before it timeouts (60s in the example).
-    #     do_something()
-    #     mutex.unlock
-    #   else
-    #     mutex.wait
-    #   end
-    # rescue => e
-    #   mark_deleting_locks
-    #   raise 'what action fails because of the lock error'
-    # end
-    #
-    # NOTE:
-    #   The difference between modified time and current time determines if timeout or not.
-    #   So the actully time to wait may be longer than the time set by `expired`.
-    #
-    class FileMutex
-      attr_reader :expired
-
-      def initialize(lock_name, logger, expired = 60)
-        @file_path = "#{CPI_LOCK_DIR}/#{lock_name}"
-        @logger = logger
-        @expired = expired
-        @is_locked = false
-      end
-
-      def lock
-        begin
-          mtime = File.mtime(@file_path)
-        rescue Errno::ENOENT => e
-          # lock file does not exist so we can try to lock
-          begin
-            fd = IO.sysopen(@file_path, Fcntl::O_WRONLY | Fcntl::O_EXCL | Fcntl::O_CREAT) # Using O_EXCL, creation fails if the file exists
-            f = IO.open(fd)
-            f.syswrite(Process.pid.to_s)
-            @logger.debug("The lock `#{@file_path}' is created by the process `#{Process.pid}'")
-            @is_locked = true
-          rescue Errno::EEXIST => e
-            @logger.info("Failed to create the lock file `#{@file_path}' because it has been created by another process.")
-            return false
-          ensure
-            f&.close
-          end
-          return true
-        end
-
-        if Time.new - mtime > @expired
-          @logger.debug("The lock `#{@file_path}' exists, but timeouts.")
-          raise LockTimeoutError
-        end
-
-        @logger.debug("The lock `#{@file_path}' exists")
-        false
-      end
-
-      def wait
-        loop do
-          begin
-            mtime = File.mtime(@file_path)
-          rescue Errno::ENOENT => e
-            @logger.debug("The lock `#{@file_path}' does not exist")
-            return true
-          end
-          raise LockTimeoutError if Time.new - mtime > @expired
-          sleep(1) # second
-        end
-      end
-
-      def unlock
-        File.delete(@file_path)
-        @logger.debug("The lock `#{@file_path}' is deleted by the process `#{Process.pid}'")
-        @is_locked = false
-      rescue Errno::ENOENT => e
-        raise LockNotFoundError
-      end
-
-      def update
-        raise LockNotOwnedError unless @is_locked
+    def flock(lock_name, mode)
+      file_path = "#{CPI_LOCK_DIR}/#{lock_name}"
+      file = File.open(file_path, File::RDWR | File::CREAT, 0o644)
+      success = file.flock(mode)
```

Comment:
I think flock does not works on some file systems, say the nfs,
if customer mount  :(
it's better to check the file system first.

and actually I think the best solution for the inter-process lock would be the azure storage blob lease.

Reply:
1. There is no any document in https://ruby-doc.org/core-2.4.2/File.html#method-i-flock saying that flock is not supported by NFS. And I just had a simple test in my local ubuntu VM, it works well when the file is in a NFS folder.
2. Azure storage blob lease is not a good solution for this case, we need some `exclusive` and `share` locks, but blob lease is always a `exclusive` lock. Another important thing is that we will use the lock in creating storage account, but in some case we can use the blob lease as the storage account does not exist at that time.
3. Won't consider named pipe this time, as flock is good enough.

Reply:
https://github.com/rubygems/rubygems/pull/1327

seems the NFS support is added.